### ISSUE TITLE
CAM/Scaling replaces M commands

### ISSUE THREAD

**Issue by smooker** [vlachoudis/bCNC#1200]
Replaces M04 GRBL laser command with M03 during scale.

**Comment by Harvie** [comment]
@MARIOBASZ any ideas?

**Comment by smooker** [comment]
Additional info:
vim: build/lib/bCNC/plugins/scaling.py 209,24-59

```
                                        bid_block.append("(---- Scale (x "+str(xscale)+" : 1.0),(y "+str(yscale)+" : 1.0),(z "+str(zscale)+" : 1.0) ---- )")
                                        bid_block.append("(center "+str(center[0])+" ,"+str(center[1])+" )")
                                        bid_block.append("M03")
                                        bid_block.append("S "+str(rpm))
                                        bid_block.append(CNC.zsafe())
                                        bid_block.append("F "+str(zfeed))
                                        bid_block.append("g0 x "+str(info[0])+" y "+str(info[1]))
```

CNC.zsafe() is not needed here for laser engraving. (milling of scaled model too :) slows down the process or breaks the instruments with coefficients bigger than 1.000)
A lot of lower cases and spaces.

Regards,
smooker.

**Comment by MARIOBASZ** [comment]
Never use laser.
I will add Laser option.
What should be on the Gcode replacing Zsafe?
Must say M03 activate, M05 deactivate?
It is easy to modify in the plugin.

**Comment by MARIOBASZ** [comment]
& Harvie, Laser option is registered as variable in bcnc? If it is correct, you tell me how to summon it thus saving time, and it is not necessary to place the option in the plugin?

**Comment by smooker** [comment]
M03 or M04 in front of every path (block)
M05 to turn off the laser in the end of the path (to not make scratches between bridges :) )

**Comment by MARIOBASZ** [comment]
M03 or M04 in front of every path (block)
M05 to turn off the laser in the end of the path (to not make scratches between bridges :) )
Ok

**Comment by MARIOBASZ** [comment]
@smooker, I did not understand what are you saying:

I think that Zsafe should be scaled too (in milling mode) and ingnored in laser.
Opinions ?

**Comment by MARIOBASZ** [comment]
place the laser / milling option
if there is a "laser variable" stored in bcnc, I will try to call it (@Harvie, delete an earlier comment by mistake, can you tell me if it is stored in bcnc as a variable, and how do I call a laser variable to save time?)
If it does not exist, in the plugin I will place the Laser / engraving option

**Comment by Harvie** [comment]
Scaling plugin should only change the scale of g-code coordinates. I don't understand why it should change how motors/lasers are engaged...

**Comment by MARIOBASZ** [comment]
I'm not in my machine.
but at the end of the block must go to zsafe for safety in milling, and for laser they say "change between m03 / m05)

**Comment by MARIOBASZ** [comment]
@smooker, if you send me a simple code to burn a square with laser, or two squares, one in each block, I will understand faster

**Comment by smooker** [comment]
just a minute

**Comment by MARIOBASZ** [comment]
@Harvie, when you do laser pocket operation, in more than one figure, write m03m05 instead of zsafe ztarget?

**Comment by MARIOBASZ** [comment]
@smooker, I'm glad you can use scaling.
here a use to scaling and modified (even in test) for 3d, oval from curculus.
Disadvantage: My wife told me "Mario, go and take a shower"
![49F47F67-141F-4F4E-8CBE-37B966E9C874](https://user-images.githubusercontent.com/11743403/54879134-55f5df00-4e14-11e9-8f23-b28687c9cfcc.jpeg)

**Comment by smooker** [comment]
smooker@sh21 /mnt/vg3.2/home/smooker/src/bCNC/examples $ cat MARIOBASZ.ngc 
```
G21 (All units in mm)

S1000 (speed of the motor is pwm to laser)

G00 X0.0 Y0.0 Z0.0 (laser stays on Z=0 always)
M04 (turn on grbl last in "auto" mode)
G01 X0.0 Y10.0 Z0.000000 F400.000000
G01 X10.0 Y10.0 Z0.000000
G01 X10.0 Y0.0 Z0.000000
G01 X0.0 Y0.0 Z0.000000
M05

G00 X20.0 Y0.0
G91
M04 (turn on grbl last in "auto" mode)
G01 X0.0 Y10.0 Z0.000000 F400.000000
G01 X10.0 Y0.0 Z0.000000
G01 X0.0 Y-10.0 Z0.000000
G01 X-10.0 Y0.0 Z0.000000
M05
```

**Comment by MARIOBASZ** [comment]
Thank you!
Have you corrected the plugin? Do you want to make your modification? Do you want me to do it?

**Comment by smooker** [comment]
I've tried, but w/o success. Have no idea of python :)
It's enough just to bypass the M commands as they are and as Harvie said :)

**Comment by smooker** [comment]
![Screenshot_2019-03-24_12-39-49](https://user-images.githubusercontent.com/5567758/54879354-a6833180-4e40-11e9-9664-10f2a06f2123.png)
![Screenshot_2019-03-24_12-40-17](https://user-images.githubusercontent.com/5567758/54879355-a84cf500-4e40-11e9-80e6-5a12605aa9b0.png)
![Screenshot_2019-03-24_12-40-39](https://user-images.githubusercontent.com/5567758/54879357-aaaf4f00-4e40-11e9-9a51-59c580ce9b38.png)

**Comment by MARIOBASZ** [comment]
it is that it selects the route, and on it realizes scale, I think that it ignores the line that says m04 m05, it should fix me. At least I will put "select for laser" if it is simpler, until final solution.
One question, for when installing a laser: ignore height z, or if you write z-1 low z?

**Comment by smooker** [comment]
Do not ignore Z, because some people may use it for focus/defocus (increase/decrease the area of the "tool").
Just scale Z too. It'll be up to the user to put the correct scale factor.

**Comment by MARIOBASZ** [comment]
with laser you use pocket operation? automatically type m04 m05?

**Comment by smooker** [comment]
yes. it's not needed to turn off the laser with M05 for deeper engraving (pocket), but it's ok.

**Comment by MARIOBASZ** [comment]
I see that you are trying to modify the plugin, Animo! Believe me I barely got some python!
experience modifying what I wrote, surely you will be able to do it, but between today and tomorrow I modify it so that it serves you and you do not have to correct the gcode by hand

**Comment by MARIOBASZ** [comment]
Place boolean laser field (look at trochoid 3d plugin or helical, I think there is a Boolean field)
At the end you must say Zsafe, for no laser and m05 for laser

**Comment by MARIOBASZ** [comment]
How BCNC behaves profile, pocket or cut, to know if it is with laser or with milling? (write M03 M05 or write Z safe or Z objective?)
Is there any laser / cutter option?

**Comment by MARIOBASZ** [comment]
I have added an option for Laser scaling (as long as you do not know how bcnc knows that you should write for laser, it is a solution)
@smooker, can you check that everything works correctly?

**Comment by smooker** [comment]
sorry for the delay. will check later and post the result

**Comment by smooker** [comment]
I can not see any difference :)
Which one is the commit ?

**Comment by MARIOBASZ** [comment]
[scaling.zip](https://github.com/vlachoudis/bCNC/files/3025746/scaling.zip)

**Comment by smooker** [comment]
![Screenshot_2019-03-30_19-28-07](https://user-images.githubusercontent.com/5567758/55279546-40a41780-5322-11e9-966c-6e0aaf412ef7.png)
![Screenshot_2019-03-30_19-28-30](https://user-images.githubusercontent.com/5567758/55279547-40a41780-5322-11e9-8fc0-d9ab244a256a.png)
![Screenshot_2019-03-30_19-29-09](https://user-images.githubusercontent.com/5567758/55279548-413cae00-5322-11e9-868c-4a24292f6c1f.png)

**Comment by smooker** [comment]
![Screenshot_2019-03-30_19-32-22](https://user-images.githubusercontent.com/5567758/55279574-9973b000-5322-11e9-9803-6f72c0fcc825.png)
Just a suggestion...
